File: packages/provide-uterm-go/serverconfig/profiles_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package serverconfig

import (
	"testing"
)

func sampleProfile(id, owner, visibility string) ConnectionProfile {
	return ConnectionProfile{
		ProfileID: id, Owner: owner, Name: "P", ConnectorType: "ssh",
		Visibility: visibility, CreatedAt: 1, UpdatedAt: 1,
	}
}

func TestProfileStoreCRUD(t *testing.T) {
	store := NewFileProfileStore(t.TempDir())

	if _, err := store.CreateProfile(sampleProfile("p1", "alice", "private")); err != nil {
		t.Fatal(err)
	}
	if _, err := store.CreateProfile(sampleProfile("p2", "bob", "shared")); err != nil {
		t.Fatal(err)
	}

	got, err := store.GetProfile("p1")
	if err != nil || got == nil || got.Owner != "alice" {
		t.Fatalf("get p1: %v %+v", err, got)
	}
	missing, err := store.GetProfile("nope")
	if err != nil || missing != nil {
		t.Fatalf("get missing: %v %+v", err, missing)
	}

	alice := "alice"
	visible, err := store.ListProfiles(&alice)
	if err != nil {
		t.Fatal(err)
	}
	// alice sees her own (p1) + shared (p2).
	if len(visible) != 2 {
		t.Errorf("alice visible = %d, want 2", len(visible))
	}

	all, err := store.ListProfiles(nil)
	if err != nil || len(all) != 2 {
		t.Fatalf("list all: %v len=%d", err, len(all))
	}

	// Update only mutable fields; connector_type (immutable) is ignored.
	updated, err := store.UpdateProfile("p1", map[string]any{"name": "Renamed", "connector_type": "telnet"})
	if err != nil || updated == nil {
		t.Fatalf("update: %v %+v", err, updated)
	}
	if updated.Name != "Renamed" || updated.ConnectorType != "ssh" {
		t.Errorf("update applied wrong fields: %+v", updated)
	}
	if updated.UpdatedAt == 1 {
		t.Errorf("updated_at not refreshed")
	}

	if noUpdate, err := store.UpdateProfile("nope", map[string]any{"name": "x"}); err != nil || noUpdate != nil {
		t.Fatalf("update missing: %v %+v", err, noUpdate)
	}

	ok, err := store.DeleteProfile("p1")
	if err != nil || !ok {
		t.Fatalf("delete p1: %v %v", err, ok)
	}
	ok2, err := store.DeleteProfile("p1")
	if err != nil || ok2 {
		t.Fatalf("second delete should be false: %v %v", err, ok2)
	}
}

func TestProfileValidateRejectsBadConnectorType(t *testing.T) {
	store := NewFileProfileStore(t.TempDir())
	bad := sampleProfile("p", "a", "private")
	bad.ConnectorType = "bogus"
	if _, err := store.CreateProfile(bad); err == nil {
		t.Fatal("expected connector_type validation error")
	}
}

func TestProfileEmptyStoreReturnsEmpty(t *testing.T) {
	store := NewFileProfileStore(t.TempDir())
	all, err := store.ListProfiles(nil)
	if err != nil || len(all) != 0 {
		t.Fatalf("empty store: %v len=%d", err, len(all))
	}
}

File: packages/provide-uterm-ts/src/serverconfig/profiles.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

import { describe, expect, it } from "vitest";
import { loadGolden } from "../testing/golden.ts";
import {
  applyProfileUpdate,
  type ConnectionProfile,
  findProfile,
  MUTABLE_PROFILE_FIELDS,
  removeProfile,
  visibleProfiles,
} from "./index.ts";

interface ProfilesGolden {
  mutable_fields: string[];
  profiles: ConnectionProfile[];
  listings: Array<{ owner: string | null; visible: string[] }>;
  updates: Array<{ name: string; updates: Record<string, unknown>; result: ConnectionProfile }>;
  updated_at: number;
}

const golden = loadGolden<ProfilesGolden>("profiles_golden.json");

/** The first recorded profile, which the update cases are applied to. */
const subject = golden.profiles[0] as ConnectionProfile;

describe("which profiles a caller sees", () => {
  it.each(golden.listings)("owner $owner", (record) => {
    expect(visibleProfiles(golden.profiles, record.owner ?? undefined).map((p) => p.profile_id)).toEqual(
      record.visible,
    );
  });

  it("shows a caller their own and the shared ones", () => {
    const seen = visibleProfiles(golden.profiles, "alice").map((p) => p.profile_id);
    expect(seen).toContain("p1");
    expect(seen).toContain("p2");
    expect(seen).toContain("p4");
  });

  it("never shows another person's private target", () => {
    expect(visibleProfiles(golden.profiles, "alice").map((p) => p.profile_id)).not.toContain("p3");
    expect(visibleProfiles(golden.profiles, "carol").map((p) => p.profile_id)).toEqual(["p2", "p4"]);
  });

  it("shows everything to a caller with no owner at all", () => {
    // The administrative view, and why a caller's identity has to reach this
    // rather than being assumed.
    expect(visibleProfiles(golden.profiles).map((p) => p.profile_id)).toEqual(golden.profiles.map((p) => p.profile_id));
  });

  it("treats an empty owner as somebody, not as nobody", () => {
    // A caller identified as the empty string owns nothing; reading it as
    // "no owner given" would show them every private profile there is.
    expect(visibleProfiles(golden.profiles, "").map((p) => p.profile_id)).toEqual(["p2", "p4"]);
  });

  it("does not hand back the list it was given", () => {
    const listed = visibleProfiles(golden.profiles);
    expect(listed).not.toBe(golden.profiles);
  });
});

describe("what an update may change", () => {
  it.each(golden.updates)("$name", (record) => {
    expect(applyProfileUpdate(subject, record.updates, golden.updated_at)).toEqual(record.result);
  });

  it("names the fields the reference names", () => {
    expect([...MUTABLE_PROFILE_FIELDS].sort()).toEqual(golden.mutable_fields);
  });

  it("refuses to move a profile to another owner", () => {
    // The point of the allow-list: a client that could rewrite an owner could
    // hand itself somebody else's saved target by editing a profile rather
    // than by asking for it.
    expect(applyProfileUpdate(subject, { owner: "mallory" }, golden.updated_at).owner).toBe(subject.owner);
  });

  it("refuses to change what identifies a profile", () => {
    const updated = applyProfileUpdate(
      subject,
      { profile_id: "stolen", connector_type: "shell", created_at: 0 },
      golden.updated_at,
    );
    expect(updated.profile_id).toBe(subject.profile_id);
    expect(updated.connector_type).toBe(subject.connector_type);
    expect(updated.created_at).toBe(subject.created_at);
  });

  it("keeps a permitted change alongside a forbidden one", () => {
    // Dropped rather than refused: a round-tripped profile carries every
    // field, so refusing the whole request would make the ordinary case fail.
    const updated = applyProfileUpdate(subject, { name: "renamed", owner: "mallory" }, golden.updated_at);
    expect(updated.name).toBe("renamed");
    expect(updated.owner).toBe(subject.owner);
  });

  it("ignores a field nobody defined", () => {
    expect(applyProfileUpdate(subject, { nonsense: 1 }, golden.updated_at)).toEqual(
      applyProfileUpdate(subject, {}, golden.updated_at),
    );
  });

  it("stamps the time itself rather than taking the caller's", () => {
    // So it records when the change happened, not what a client claimed.
    expect(applyProfileUpdate(subject, { updated_at: 0 }, golden.updated_at).updated_at).toBe(golden.updated_at);
    expect(applyProfileUpdate(subject, {}, 42).updated_at).toBe(42);
  });

  it("does not change the profile it was given", () => {
    const before = JSON.stringify(subject);
    applyProfileUpdate(subject, { name: "renamed" }, golden.updated_at);
    expect(JSON.stringify(subject)).toBe(before);
  });
});

describe("finding and removing a profile", () => {
  it("finds one by its identifier", () => {
    expect(findProfile(golden.profiles, "p3")?.owner).toBe("bob");
    expect(findProfile(golden.profiles, "nope")).toBeUndefined();
  });

  it("removes one and says it did", () => {
    const { profiles, removed } = removeProfile(golden.profiles, "p1");
    expect(removed).toBe(true);
    expect(profiles.map((p) => p.profile_id)).toEqual(["p2", "p3", "p4"]);
  });

  it("says so when there was nothing to remove", () => {
    // A caller can tell a deletion from a request to delete something that
    // never existed.
    const { profiles, removed } = removeProfile(golden.profiles, "nope");
    expect(removed).toBe(false);
    expect(profiles).toHaveLength(golden.profiles.length);
  });

  it("does not change the list it was given", () => {
    removeProfile(golden.profiles, "p1");
    expect(golden.profiles).toHaveLength(4);
  });
});
